pkg/cache/v2/linear.go

```diff
@@ -0,0 +1,219 @@
+}
+
+var _ Cache = &LinearCache{}
+
+func NewLinearCache(typeURL string, initialResources map[string]types.Resource) *LinearCache {
+	if initialResources == nil {
+		initialResources = make(map[string]types.Resource)
+	}
+	out := &LinearCache{
+		typeURL:       typeURL,
+		resources:     initialResources,
+		watches:       make(map[string]watches),
+		watchAll:      make(watches),
+		version:       0,
+		versionVector: make(map[string]uint64),
+	}
+	for name := range initialResources {
+		out.versionVector[name] = 0
+	}
+	return out
+}
+
+func (cache *LinearCache) respond(value chan Response, staleResources []string) {
+	var resources []types.Resource
+	// TODO: optimize the resources slice creations across different clients
+	if len(staleResources) == 0 {
+		resources = make([]types.Resource, 0, len(cache.resources))
+		for _, resource := range cache.resources {
+			resources = append(resources, resource)
+		}
+	} else {
+		resources = make([]types.Resource, 0, len(staleResources))
+		for _, name := range staleResources {
+			resource := cache.resources[name]
+			if resource != nil {
+				resources = append(resources, resource)
+			}
+		}
+	}
+	value <- RawResponse{
+		Request:   Request{TypeUrl: cache.typeURL},
+		Resources: resources,
+		Version:   strconv.FormatUint(cache.version, 10),
+	}
+}
+
+func (cache *LinearCache) notifyAll(modified map[string]struct{}) {
+	// de-duplicate watches that need to be responded
+	notifyList := make(map[chan Response][]string)
+	for name := range modified {
+		for watch := range cache.watches[name] {
+			notifyList[watch] = append(notifyList[watch], name)
+		}
+		delete(cache.watches, name)
+	}
+	for value, stale := range notifyList {
+		cache.respond(value, stale)
+	}
+	for value := range cache.watchAll {
+		cache.respond(value, nil)
+	}
+	cache.watchAll = make(watches)
+}
+
+// UpdateResource updates a resource in the collection.
+func (cache *LinearCache) UpdateResource(name string, res types.Resource) error {
+	if res == nil {
+		return errors.New("nil resource")
+	}
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	cache.version += 1
+	cache.versionVector[name] = cache.version
+	cache.resources[name] = res
+
+	// TODO: batch watch closures to prevent rapid updates
+	cache.notifyAll(map[string]struct{}{name: struct{}{}})
+
+	return nil
+}
+
+// DeleteResource removes a resource in the collection.
+func (cache *LinearCache) DeleteResource(name string) error {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	cache.version += 1
+	delete(cache.versionVector, name)
+	delete(cache.resources, name)
+
+	// TODO: batch watch closures to prevent rapid updates
+	cache.notifyAll(map[string]struct{}{name: struct{}{}})
+	return nil
+}
+
+func (cache *LinearCache) CreateWatch(request Request) (chan Response, func()) {
+	value := make(chan Response, 1)
+	if request.TypeUrl != cache.typeURL {
+		value <- RawResponse{Request: Request{TypeUrl: cache.typeURL}, Version: strconv.FormatUint(cache.version, 10)}
```

Comment:
Should we be closing the channel here as well? It's unclear to me what implications there are to the receiver if we send back a response with a different type url than the original request.

Reply:
Good point. Returning a closed channel makes sense here, too.